File: bindings/typescript/src/feedback.ts
/**
 * User feedback on an answer (PR-4.1) — a 👍/👎 rating + optional note the
 * gateway records into its rebuildable-to-empty `feedback.db` sidecar via
 * `SubmitFeedback`, read back by `ListFeedback`. AUDIT/DISPLAY ONLY: client-origin
 * product signal, never truth, never identity, never a digest input. Kept in its
 * own module (the telemetry.ts module-per-concern precedent).
 *
 * SN-8: the caller principal + the `feedbackId` are server-derived; the SDK only
 * *encodes* the bytes to hex and maps the rating enum.
 */

import type { FeedbackRow as PbFeedbackRow } from "./gen/kortecx/v1/gateway_pb.js";
import { FeedbackRating } from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** A 👍/👎 rating as a stable string (the wire enum, lowercased). */
export type Rating = "up" | "down";

/** Map a {@link Rating} string to the proto enum int (UP=1, DOWN=2). */
export function ratingToProto(rating: Rating): FeedbackRating {
  return rating === "up" ? FeedbackRating.UP : FeedbackRating.DOWN;
}

/** Map the proto rating enum int back to a {@link Rating} (`null` if unset). */
export function ratingFromProto(r: FeedbackRating): Rating | null {
  if (r === FeedbackRating.UP) return "up";
  if (r === FeedbackRating.DOWN) return "down";
  return null;
}

/** The target + context the UI supplies when rating an answer. `messageId` is the
 *  stable per-answer key (required); the rest are advisory join/context fields
 *  (all-zero/absent is fine for a local-only turn with no run). */
export interface FeedbackInput {
  readonly rating: Rating;
  readonly messageId: string;
  readonly instanceId?: string;
  readonly moteId?: string;
  readonly contentRef?: string;
  readonly comment?: string;
  readonly recipeHandle?: string;
  readonly modelId?: string;
}

/** One recorded feedback row in a {@link FeedbackPage} (newest-first). `instanceId`
 *  is `""` when the turn had no run; `moteId`/`contentRef` are `""` when absent. */
export class FeedbackRow {
  constructor(
    readonly feedbackId: string,
    readonly rating: Rating | null,
    readonly messageId: string,
    readonly instanceId: string,
    readonly moteId: string,
    readonly contentRef: string,
    readonly comment: string,
    readonly recipeHandle: string,
    readonly modelId: string,
    readonly submittedUnixMs: number,
    readonly rowid: number,
  ) {}

  static fromProto(r: PbFeedbackRow): FeedbackRow {
    const nonEmpty = (b: Uint8Array): string => (b.some((x) => x !== 0) ? encode(b) : "");
    return new FeedbackRow(
      encode(r.feedbackId),
      ratingFromProto(r.rating),
      r.messageId,
      nonEmpty(r.instanceId),
      nonEmpty(r.moteId),
      nonEmpty(r.contentRef),
      r.comment,
      r.recipeHandle,
      r.modelId,
      Number(r.submittedUnixMs),
      Number(r.rowid),
    );
  }

  /** A plain snake_case object (stable wire-shaped serialization for UIs/logs). */
  toJSON() {
    return {
      feedback_id: this.feedbackId,
      rating: this.rating,
      message_id: this.messageId,
      instance_id: this.instanceId,
      mote_id: this.moteId,
      content_ref: this.contentRef,
      comment: this.comment,
      recipe_handle: this.recipeHandle,
      model_id: this.modelId,
      submitted_unix_ms: this.submittedUnixMs,
      rowid: this.rowid,
    };
  }
}

/** One page of {@link FeedbackRow} (newest-first) plus the `hasMore` cursor flag. */
export interface FeedbackPage {
  readonly rows: FeedbackRow[];
  readonly hasMore: boolean;
}

File: bindings/python/src/kortecx/feedback.py
"""User feedback on an answer (PR-4.1) — a 👍/👎 rating + optional note recorded
by ``SubmitFeedback`` and read back by ``ListFeedback``.

Client-origin product signal the gateway records into a rebuildable-to-empty
``feedback.db`` sidecar — AUDIT/DISPLAY ONLY, never truth, never identity, never
a digest input. Kept in its own module (the telemetry.py / module-per-concern
precedent). SN-8: the caller principal + the ``feedback_id`` are server-derived;
the SDK only hex-encodes the bytes + maps the rating enum.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from . import hexids
from .v1 import gateway_pb2 as _g

#: A 👍/👎 rating as a stable string (the wire enum, lowercased).
Rating = str  # Literal["up", "down"]


def rating_to_proto(rating: str) -> "_g.FeedbackRating":
    """Map a :data:`Rating` string to the proto rating enum (UP=1, DOWN=2)."""
    if rating == "up":
        return _g.FEEDBACK_RATING_UP
    if rating == "down":
        return _g.FEEDBACK_RATING_DOWN
    raise ValueError(f"rating must be 'up' or 'down', got {rating!r}")


def rating_from_proto(value: int) -> Optional[str]:
    """Map the proto rating enum int back to a :data:`Rating` (``None`` if unset)."""
    if value == _g.FEEDBACK_RATING_UP:
        return "up"
    if value == _g.FEEDBACK_RATING_DOWN:
        return "down"
    return None


@dataclass(frozen=True)
class FeedbackRow:
    """One recorded feedback row. ``instance_id`` is all-zero hex (``""`` after
    normalization) when the turn had no run; ``mote_id``/``content_ref`` are
    ``""`` when absent. ``rowid`` is the pagination cursor (never identity)."""

    feedback_id: str  # hex, server-derived
    rating: Optional[str]  # "up" | "down" | None
    message_id: str
    instance_id: str  # hex ("" when no run)
    mote_id: str  # hex ("" when absent)
    content_ref: str  # hex ("" when absent)
    comment: str
    recipe_handle: str
    model_id: str
    submitted_unix_ms: int  # audit-only wall clock (off every hash)
    rowid: int

    @classmethod
    def from_proto(cls, r: "_g.FeedbackRow") -> "FeedbackRow":
        def non_empty(b: bytes) -> str:
            return hexids.encode(b) if any(b) else ""

        return cls(
            feedback_id=hexids.encode(r.feedback_id),
            rating=rating_from_proto(r.rating),
            message_id=r.message_id,
            instance_id=non_empty(r.instance_id),
            mote_id=non_empty(r.mote_id),
            content_ref=non_empty(r.content_ref),
            comment=r.comment,
            recipe_handle=r.recipe_handle,
            model_id=r.model_id,
            submitted_unix_ms=r.submitted_unix_ms,
            rowid=r.rowid,
        )


@dataclass(frozen=True)
class FeedbackPage:
    """One newest-first page of :class:`FeedbackRow` plus the ``has_more`` flag."""

    rows: List[FeedbackRow]
    has_more: bool
